### ISSUE TITLE
[Feature]: Add SpectralQuant KV Cache Compression (builds on TurboQuant #38479)

### ISSUE THREAD

**Issue by gapsong** [vllm-project/vllm#43475]
## 🚀 The feature, motivation and pitch

**SpectralQuant** is a KV cache compression method that improves on TurboQuant
(merged in #38479) by exploiting a universal structural property of transformer
attention: across six models in four architecture families, key vectors
concentrate signal in only **3–4% of the head dimension**.

By identifying these dimensions via a one-time 15-second calibration (PCA on a
small sample), SpectralQuant applies error correction only where it matters and
skips it on the remaining ~96% noise dimensions — achieving better quality *and*
better compression simultaneously.

**Paper:** "3% Is All You Need: Breaking TurboQuant's Compression Limit via
Spectral Structure" — submitted to NeurIPS 2026, arXiv pending.
Code: https://github.com/Dynamis-Labs/spectralquant

### Headline results (from paper, on NVIDIA B200)

| Metric | SpectralQuant | TurboQuant | Delta |
|---|---|---|---|
| Cosine similarity (Qwen 2.5-14B) | 0.9485 | 0.9226 | +2.59 pp |
| Compression ratio | 5.95× | 5.02× | +18.6% |
| Latency (512 tokens) | 0.257 ms/step | 0.566 ms/step | 2.2× faster |
| Perplexity (Qwen 7B) | 7.51 | 7.51 | neutral |

### Key findings

1. **Universal low-rank structure** — d_eff/head_dim ≈ 3–4% holds across Qwen
   (1.5B, 7B, 14B), Llama 3.1-8B, Mistral 7B, and Gemma 2-9B.
2. **KV asymmetry** — Keys: d_eff ≈ 4. Values: d_eff ≈ 40–55. This is why
   low-rank compression fails for values but succeeds for keys.
3. **Statistically significant** — 10-seed Wilcoxon p=0.031 on Qwen 2.5-1.5B.

## 📋 Current status

I am currently **reproducing the results on an RTX 4090** before proposing any
implementation. I want to verify the core cosine similarity improvement holds on
consumer hardware before going further.

I'll update this issue with:
- [ ] Reproduction results on RTX 4090 (cosine sim SQ vs TQ, d_eff ratio observed)
- [ ] arXiv link once available
- [ ] Draft implementation design (calibration storage, Triton kernel approach)

## 🔧 Proposed integration

SpectralQuant builds directly on the TurboQuant backend (#38479). The main
design questions I'd like early feedback on:

**1. Calibration storage** — SpectralQuant requires a one-time eigenbasis
calibration per model. Where should this live in vLLM's loading pipeline?
Options: alongside model weights (like AWQ scales), or as a separate
`--spectralquant-calibration-path` flag.

**2. New dtype flag** — `--kv-cache-dtype spectralquant` following the same
pattern as `turboquant_k8v4`.

**3. Kernel approach** — Selective QJL correction on signal dims only,
fused into the existing TurboQuant Triton store kernel.

## ❓ Questions for maintainers

1. Is this the right place to track this, or would a Discussions thread be
   preferred while the paper is still under review?
2. Any prior art on calibration data stored alongside model weights in vLLM
   I should look at as a reference?
3. Any concerns with the calibration requirement given vLLM's preference for
   stateless compression at serve time?

## 📎 Related
- #38479 — TurboQuant merged PR (direct dependency)
- #38171 — Original TurboQuant feature request  
- TurboQuant paper: https://arxiv.org/abs/2504.19874
- SpectralQuant code: https://github.com/Dynamis-Labs/spectralquant
- Author's overview (LinkedIn): https://www.linkedin.com/pulse/3-all-you-need-breaking-turboquants-compression-limit-ashwin-gopinath-7jzkc/
  *(arXiv link will replace this once available)*

@niashwin

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by jshaofa-ui** [comment]
# Solution: vLLM #43475 — SpectralQuant KV Cache Compression

## Issue Summary
**GitHub**: vllm-project/vllm #43475
**Title**: Feature request — KV cache compression using SpectralQuant technique
**Labels**: feature request
**Competition**: Zero (0 comments, 0 PRs)
**Quote**: $2,000-$3,000

## Root Cause Analysis

### Problem Context: KV Cache Memory Bottleneck

As LLM context windows grow (8K → 32K → 128K+ tokens), the KV cache becomes the dominant memory consumer during inference. For a 70B model with 128K context:
- KV cache can consume 40-60GB per request on FP16
- This limits batch size and increases cost per request
- Memory bandwidth for KV cache reads is the primary throughput bottleneck

### Existing Approaches and Their Limitations

| Technique | Memory Savings | Quality Impact | Complexity |
|-----------|---------------|----------------|------------|
| PagedAttention (vLLM built-in) | Baseline | None | Low |
| KV quantization (INT8/INT4) | 2-4x | Moderate quality loss | Medium |
| Multi-Query Attention (MQA) | ~2x | Some quality loss | Low (arch change) |
| Grouped-Query Attention (GQA) | ~1.5x | Minimal quality loss | Low (arch change) |
| Token eviction (H2O, SnapKV) | Variable | Depends on eviction policy | Medium |
| **SpectralQuant (proposed)** | **3-5x** | **Minimal quality loss** | **High** |

### Why SpectralQuant?

SpectralQuant leverages the observation that KV cache tensors have low effective rank in the spectral domain. By projecting KV values onto a learned spectral basis and quantizing the coefficients, it achieves:

1. **Higher compression ratios** than naive per-tensor quantization (3-5x vs 2-4x)
2. **Lower quality degradation** because the spectral basis preserves the most important signal components
3. **Adaptive precision** — different attention heads and layers can use different compression levels

### Technical Foundation

The SpectralQuant technique works as follows:

1. **Spectral Decomposition**: Apply DCT (Discrete Cosine Transform) or learned spectral basis to KV cache tensors along the sequence dimension
2. **Coefficient Quantization**: Quantize the spectral coefficients using non-uniform quantization (more bits for low-frequency components, fewer for high-frequency)
3. **Adaptive Compression**: Dynamically adjust compression ratio based on attention head importance and layer depth

## Proposed Fix

### Fix 1: Spectral Transform Module

**File**: `vllm/attention/spectral_quant.py`

```python
"""
SpectralQuant: KV Cache Compression via Spectral Decomposition

This module implements SpectralQuant-based KV cache compression that achieves
3-5x memory savings with minimal quality degradation by exploiting the
low-rank structure of KV cache in the spectral domain.
"""

import torch
import torch.nn as nn
import torch.nn.functional as F
from typing import Optional, Tuple, Dict, List
from dataclasses import dataclass, field
import math


@dataclass
class SpectralQuantConfig:
    """Configuration for SpectralQuant compression."""
    # Compression settings
    compression_ratio: float = 4.0  # Target compression ratio
    min_bits: int = 2  # Minimum bits per coefficient
    max_bits: int = 8  # Maximum bits per coefficient
    
    # Spectral basis settings
    basis_type: str = "dct"  # "dct" or "learned"
    num_basis_vectors: Optional[int] = None  # None = auto from compression_ratio

**Comment by Milan933-coder** [comment]
Hey I guys Can We work Upon this Together @gapsong and @jshaofa-ui

**Comment by gapsong** [comment]
@Milan933-coder I think @jshaofa-ui is a bot, but sure lets work together on that. Do you have a GPU to run this algorithm to check the benchmarks?

**Comment by Milan933-coder** [comment]
Modal provides free GPU service they initially give 30 dollar compute after card verification for free.and let me read the spectralquant research paper first @gapsong

**Comment by jagmarques** [comment]
For the KV compression landscape: we have an E8 lattice VQ method (NexusQuant) in the same sub-4-bit space. Calibration-free, Hadamard rotation. On Llama-3.1-8B-Instruct 4K (n=30 NIAH): 30/30 at 2.0 bpe vs TurboQuant 0/30 at 2.125 bpe (McNemar b=29/c=0). Draft vLLM PR: #47742. Demo: huggingface.co/spaces/jmarquex/nexusquant-demo

**Comment by samuelsalfati** [comment]
Some research findings that might be useful for evaluating this proposal.

The spectral premise here matches what I found benchmarking KV compression at long context: concentrating signal in a data-driven basis before compressing appears to be what makes low-bit KV survivable. Most published evals stop at ≤32K, so I specifically tested the regime beyond that — three-arm comparisons (fp16 / fp8 / compressed) at 128K context under multi-user concurrency. On Qwen3-4B, the compressed arm sustained 6 concurrent users at 128K each on one GPU at ~2.1× fp16 decode per user, with needle retrieval intact across all arms (189/189). Setup and repro scripts (my own work, so factor that in): https://huggingface.co/fraQtl/qwen3-4b-instruct-2507-kv-sidecars

The most important methodological finding, and probably the most relevant one for this thread: perplexity and short-context NIAH can both pass while retrieval quietly collapses past 64K. Gating candidates on long-context retrieval at the target length killed more of my own approaches than any other test — I'd apply that gate to any method here, including mine.

If useful, I'm glad to run SpectralQuant through the same rig for an apples-to-apples comparison.
